corehq/apps/reports/commtrack/util.py

```diff
@@ -1,4 +1,81 @@
+from django.core.exceptions import ObjectDoesNotExist
+from casexml.apps.stock.consumption import ConsumptionHelper
 from corehq.apps.locations.models import SQLLocation
+from corehq.apps.products.models import SQLProduct
+from dimagi.ext import jsonobject
+
+
+class StockLedgerValueWrapper(jsonobject.JsonObject):
+    """
+    Wrapper class to abstract StockState and the equivalent model coming out of Elasticsearch
+    """
+    domain = jsonobject.StringProperty()
+    case_id = jsonobject.StringProperty()
+    section_id = jsonobject.StringProperty()
+    entry_id = jsonobject.StringProperty()
+    balance = jsonobject.DecimalProperty()  # todo: should this be an int?
+    last_modified = jsonobject.DateTimeProperty()
+    last_modified_form_id = jsonobject.StringProperty()
+    daily_consumption = jsonobject.FloatProperty()
+    location_id = jsonobject.StringProperty()
+
+    _sql_product = None
+    _sql_location = None
+
+    def __init__(self, _obj=None, sql_product=None, sql_location=None, *args, **kwargs):
+        self._sql_product = sql_product
+        self._sql_location = sql_location
+        super(StockLedgerValueWrapper, self).__init__(_obj, *args, **kwargs)
+
+    @property
+    def sql_product(self):
+        if self.entry_id and not self._sql_product:
+            try:
+                self._sql_product = SQLProduct.objects.get(domain=self.domain, product_id=self.entry_id)
+            except ObjectDoesNotExist:
+                # todo: cache this result so multiple failing calls don't keep hitting the DB
+                return None
+
+        return self._sql_product
+
+    @property
+    def sql_location(self):
+        if self.location_id and not self._sql_location:
+            try:
+                self._sql_location = SQLLocation.objects.get(domain=self.domain, location_id=self.location_id)
+            except ObjectDoesNotExist:
+                # todo: cache this result so multiple failing calls don't keep hitting the DB
+                return None
+
+        return self._sql_location
+
+    @classmethod
+    def from_stock_state(cls, stock_state):
+        return cls(
+            case_id=stock_state.case_id,
+            section_id=stock_state.section_id,
+            entry_id=stock_state.product_id,
+            balance=stock_state.stock_on_hand,
+            last_modified=stock_state.last_modified_date,
+            last_modified_form_id=stock_state.last_modified_form_id,
+            daily_consumption=stock_state.daily_consumption,
+            location_id=stock_state.sql_location.location_id if stock_state.sql_location else None,
```

Comment:
in this context it's a django model foreign key which i believe should be automatically cached. `sql_product` and `sql_location` are also memoized on this model (i didn't use the decorator because i wanted to be able to pas them in the constructor)